packages/go_router/lib/src/builder.dart

```diff
@@ -240,19 +240,41 @@ class RouteBuilder {
         _buildRecursive(context, matchList, startIndex + 1, pagePopContext,
             routerNeglect, keyToPages, shellNavigatorKey, registry);
 
-        final HeroController heroController = _goHeroCache.putIfAbsent(
+        _goHeroCache.putIfAbsent(
             shellNavigatorKey, () => _getHeroController(context));
 
+        // Build pages for preloadable navigators
+        final Map<GlobalKey<NavigatorState>, RouteMatchList>
+            preloadableNavigators =
+            route.preloadableNavigatorLocations(configuration);
```

Comment:
can all these logic to be done in part of of `StatefulNavigationShell`'s `initState` or `didUpdateWidget` or `build`? it seems weird that builder would need to know about preload....

We probably need to refactor some API for RouteBuilder though.

Reply:
It might, but I'd I have to look in to it, but we would have to expose new API on RouteBuilder.

Reply:
Refactored this logic out of `RouteBuilder`, and at the moment into `StatefulShellRoute`. The reason I didn't move it into `StatefulNavigationShell` and `initState`/`didUpdateWidget` is that I wanted the call back into `RouteBuilder` (`buildShellNavigator`) to be in the same build cycle, and thus ensuring the `GoRouterStateRegistry`, `_PagePopContext` and `_goHeroCache` are updated correctly.